### PR TITLE
XML-Assignment 

### PR DESCRIPTION
**Pull request by skrzym**
### Summary
This pull request contains my work for section 5.3 assignment on data wrangling with XML.
It contains the following items:
* **data_wrangling_xml.py** - Python code for assignment
* **XML_assignment.ipynb** - Jupyter notebook presenting the answers and thought process for the assignment
* **sliderule_dsi_xml_exercise.ipynb** - Jupyter example notebook supplied for the assignment
* **"data" directory** - XML data stored here

### CODE REVIEW

File: 5-3_data_wrangling_xml/data_wrangling_xml.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        elevation = None
+
+    # Add dictionary of answers to 'airports' list
+    airports.append({
+        'name': name,
+        'country': country,
+        'elevation': elevation
+    })
+
+df_airports = pd.DataFrame.from_records(airports).sort_values('elevation',ascending=False).reset_index(drop=True)
```
Comment (amirziai) [inline 5-3_data_wrangling_xml/data_wrangling_xml.py]: Space after commas (see PEP8)

File: 5-3_data_wrangling_xml/data_wrangling_xml.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+highest_airport = df_airports.iloc[0]
```
Comment (amirziai) [inline 5-3_data_wrangling_xml/data_wrangling_xml.py]: Newline at the end of files http://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

File: 5-3_data_wrangling_xml/data_wrangling_xml.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
+import pandas as pd
+from math import floor
+from xml.etree import ElementTree as ET
+
+
+document = ET.parse( './data/mondial_database.xml' )
```
Comment (amirziai) [inline 5-3_data_wrangling_xml/data_wrangling_xml.py]: Extra space

File: 5-3_data_wrangling_xml/data_wrangling_xml.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+# -----------------------------------------------------
+# 10 Countries with the lowest infant mortality rates
+#
+# The approach here is to go through each country and if it has an 'infant_mortality' tag,
+# store that data as a float in a dictionary with the country's name as the key.
+
+# Create a list to hold dicts containing {country, infant_mortality_rate}
+imr_list = []
```
Comment (amirziai) [inline 5-3_data_wrangling_xml/data_wrangling_xml.py]: Consider using for comprehensions

File: 5-3_data_wrangling_xml/data_wrangling_xml.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    imr_list.append({'country': name, 'infant_mortality_rate': imr})
+
+# convert list of dicts to DataFrame
+df_imr = pd.DataFrame.from_records(imr_list)
+
+# sort by population and make a top 10 df using '.head(10)'
+top_10_imr = df_imr.sort_values('infant_mortality_rate',ascending=True).head(10)
+
+# reindex top 10 df to show index as ranking from 1-10
+top_10_imr.index = range(1,11)
```
Comment (amirziai) [inline 5-3_data_wrangling_xml/data_wrangling_xml.py]: Space after comma

**Review by amirziai — COMMENTED** [review]
You can rewrite most of your loops with list comprehensions. Take a look here: https://gist.github.com/amirziai/84cb20a4e92027becd60e48fd5924472 for some hints on getting you started

**Comment by skrzym** [comment]
I've gone through and changed the data_wrangling_xml.py file to use more list comprehensions and functions vs the for loops I had.

File: 5-3_data_wrangling_xml/data_wrangling_xml.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+)
+
+
+def make_top_10(data, col_names, sort_by, asc=False):
+    df = pd.DataFrame.from_dict(data, orient='index')
+    df.reset_index(inplace=True)
+    df.columns = col_names
+    df.sort_values(sort_by, ascending=asc, inplace=True)
+    df = df.head(10)
+    df.index = range(1, 11)
```
Comment (amirziai) [inline 5-3_data_wrangling_xml/data_wrangling_xml.py:36]: Nice!

File: 5-3_data_wrangling_xml/data_wrangling_xml.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+# Generate dictionary with name:imr pairs
+imr = dict(
+    (element_find(country, 'name'),  # Extract country's name
+     float(element_find(country, 'infant_mortality')))  # Extract infant mortality rate
+    for country in document.iterfind('country')
+    if element_exists(country, 'infant_mortality')  # Filter out countries with no imr
+)
+
+
+def make_top_10(data, col_names, sort_by, asc=False):
```
Comment (amirziai) [inline 5-3_data_wrangling_xml/data_wrangling_xml.py:30]: You could parameterize the number of elements to display/return.

File: 5-3_data_wrangling_xml/data_wrangling_xml.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+# Create a DataFrame using the 'ethnicgroup_populations' list and sum the data by ethnic group
+df_ethnic_populations = pd.DataFrame.from_records(ethnicgroup_populations, columns=['ethnicgroup', 'population'])
+df_ethnic_populations = df_ethnic_populations.groupby('ethnicgroup', sort=True).agg(sum).reset_index()
+df_ethnic_populations.sort_values('population', ascending=False, inplace=True)
+
+# Create a top 10 DataFrame
+top_10_ethnic_groups = pd.concat([
+    df_ethnic_populations.loc[:, 'ethnicgroup'].head(10),
+    df_ethnic_populations.loc[:, 'population'].apply(floor).head(10)
```
Comment (amirziai) [inline 5-3_data_wrangling_xml/data_wrangling_xml.py:93]: Good job using `apply`. Would using `floor` vs. `np.floor` make a difference here?

**Review by amirziai — APPROVED** [review]
Nice job! A couple of comments to consider.

**Pull request merged**